src/sample.rs

```diff
@@ -0,0 +1,330 @@
+// Copyright 2013-2014 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Functions for sampling data
+
+use super::Rng;
+use std::collections::hash_map::HashMap;
+
+/// The `Sample` trait provides the `sample` method.
+///
+/// This is intended to be implemented for containers that:
+/// - Can be sampled in `O(amount)` time.
+/// - Whos items can be `cloned`.
+///
+/// If cloning is impossible or expensive, use `sample_ref` instead.
+pub trait Sample {
```

Comment:
Note that traits can't be implemented outside of the module where (a) the trait is defined or (b) the target type is defined. This means no one else can implement `Sample` for `std::collections` types.

Likely though, this isn't necessary: users can instead create an iterator and call sample on that. Which begs the questions: (1) why use traits instead of plain functions and (2) why not let a more general function like `sample_reservoir` take an optional length (or a length and an iterator), and use the appropriate algorithm when length is known?

Reply:
`sample_reservoir` requires iterating through the entire sequence length `N` to sample `m` samples, so is `O(N)` computationally and `O(m)` in memory.

`Sample::sample` does *not* iterate through the sequence, so is `O(m)` for both memory **and** computation.

Reply:
Just knowing the length allows O(m) computation of indices... but getting those values from the iterator is still O(N). Ok.

Reply:
Also, `sample` guarantees that the order of the returned elements are random, where as `sample_resevoir` does not.